File: frontend/src/api/ai.ts
/**
 * AI — proposal extraction and the agent chat.
 *
 * The write path is always **suggest -> human confirm -> commit**:
 *   1. `useExtractProposal()` reads an already-uploaded document and returns a
 *      `ProposalSuggestion`. It writes NO proposal.
 *   2. The broker edits the suggestion in the UI.
 *   3. `useConfirmExtraction()` commits the reviewed payload as a real proposal.
 *
 * Nothing auto-commits, and the `extraction` row (model, prompt version, raw
 * output, confidence, source document) is persisted either way for traceability.
 */
import { useMutation, useQuery, useQueryClient } from "@tanstack/react-query";
import api, { API_URL, getAccessToken } from "@/lib/api";
import { qk } from "@/api/keys";
import type {
  AgentExchange,
  AgentMessage,
  AgentScope,
  AgentThread,
  AiSummary,
  CategoryListResponse,
  CategorySpec,
  DocumentConfirmRequest,
  DocumentConfirmResponse,
  DocumentExtractionRequest,
  DocumentExtractionResponse,
  ExtractionSuggestionResponse,
  ListResponse,
  Proposal,
  ProposalConfirmRequest,
} from "@/api/types";

// --- Extraction --------------------------------------------------------------

/** SUGGEST step. Reads one document; commits nothing. */
export function useExtractProposal() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: async (payload: { document_id: number }) => {
      const { data } = await api.post<ExtractionSuggestionResponse>(
        "/ai/proposals/extract",
        payload,
      );
      return data;
    },
    onSuccess: (data) => {
      qc.setQueryData(qk.extractions.detail(data.extraction.id), data);
    },
  });
}

/** Re-read a persisted extraction and its suggestion (the audit trail). */
export function useExtraction(extractionId: number | undefined) {
  return useQuery({
    queryKey: qk.extractions.detail(extractionId ?? 0),
    enabled: !!extractionId,
    queryFn: async () => {
      const { data } = await api.get<ExtractionSuggestionResponse>(
        `/ai/extractions/${extractionId}`,
      );
      return data;
    },
  });
}

/** COMMIT step. The reviewed (possibly edited) suggestion becomes a proposal. */
export function useConfirmExtraction(extractionId: number) {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: async (payload: ProposalConfirmRequest) => {
      const { data } = await api.post<Proposal>(
        `/ai/proposals/${extractionId}/confirm`,
        payload,
      );
      return data;
    },
    onSuccess: () => {
      void qc.invalidateQueries({ queryKey: qk.proposals.all });
      void qc.invalidateQueries({ queryKey: qk.quotes.all });
      void qc.invalidateQueries({ queryKey: qk.insurers.all });
    },
  });
}

// --- Agent chat --------------------------------------------------------------

export function useAgentThreads(enabled = true) {
  return useQuery({
    queryKey: qk.agentThreads.list(),
    enabled,
    queryFn: async () => {
      const { data } = await api.get<ListResponse<AgentThread>>("/ai/threads");
      return data;
    },
  });
}

export function useCreateAgentThread() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: async (payload: {
      scope?: AgentScope;
      entity_id?: number | null;
      title?: string | null;
    }) => {
      const { data } = await api.post<AgentThread>("/ai/threads", payload);
      return data;
    },
    onSuccess: () => {
      void qc.invalidateQueries({ queryKey: qk.agentThreads.lists() });
    },
  });
}

export function useAgentMessages(threadId: number | undefined) {
  return useQuery({
    queryKey: qk.agentThreads.messages(threadId ?? 0),
    enabled: !!threadId,
    queryFn: async () => {
      const { data } = await api.get<ListResponse<AgentMessage>>(
        `/ai/threads/${threadId}/messages`,
      );
      return data;
    },
  });
}

/** One turn: posts the user message, returns it together with the answer. */
export function useSendAgentMessage(threadId: number) {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: async (payload: { content: string }) => {
      const { data } = await api.post<AgentExchange>(
        `/ai/threads/${threadId}/messages`,
        payload,
      );
      return data;
    },
    onSuccess: () => {
      void qc.invalidateQueries({ queryKey: qk.agentThreads.messages(threadId) });
      void qc.invalidateQueries({ queryKey: qk.agentThreads.lists() });
    },
  });
}

// --- Generic, registry-driven extraction --------------------------------------

/**
 * The document-category registry as JSON.
 *
 * This is what makes the extraction review form generic: a new category needs
 * a backend schema and NO new React component. Cached for the session — the
 * registry only changes when the backend redeploys.
 */
export function useCategoryRegistry(enabled = true) {
  return useQuery({
    queryKey: qk.aiCategories.lists(),
    enabled,
    staleTime: 30 * 60 * 1000,
    queryFn: async () => {
      const { data } = await api.get<CategoryListResponse>("/ai/categories");
      return data;
    },
  });
}

/** Look one category up in the registry. `undefined` = no schema, no AI button. */
export function useCategorySpec(category: string | null | undefined): CategorySpec | undefined {
  const { data } = useCategoryRegistry();
  if (!category || !data) return undefined;
  return data.items.find(
    (spec) => spec.category === category || spec.canonical_category === category,
  );
}

/** SUGGEST step for ANY category. Writes an `extraction` row, commits nothing. */
export function useExtractDocument() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: async (payload: DocumentExtractionRequest) => {
      const { data } = await api.post<DocumentExtractionResponse>(
        "/ai/documents/extract",
        payload,
      );
      return data;
    },
    onSuccess: (data) => {
      qc.setQueryData(qk.extractions.detail(data.extraction.id), data);
    },
  });
}

/** COMMIT step: the human-reviewed payload is written to its prefill target. */
export function useConfirmDocumentExtraction(extractionId: number) {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: async (payload: DocumentConfirmRequest) => {
      const { data } = await api.post<DocumentConfirmResponse>(
        `/ai/documents/${extractionId}/confirm`,
        payload,
      );
      return data;
    },
    onSuccess: () => {
      // A confirmation can touch almost any module, so invalidate broadly.
      void qc.invalidateQueries({ queryKey: qk.proposals.all });
      void qc.invalidateQueries({ queryKey: qk.caseFiles.all });
      void qc.invalidateQueries({ queryKey: qk.documents.all });
    },
  });
}

// --- Summaries (prose is a suggestion too) -------------------------------------

export function useSummarizeProposal(proposalId: number) {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: async () => {
      const { data } = await api.post<AiSummary>(`/ai/proposals/${proposalId}/summary`);
      return data;
    },
    onSuccess: () => {
      void qc.invalidateQueries({ queryKey: qk.proposals.detail(proposalId) });
    },
  });
}

export function useSummarizeCaseFile(caseId: number) {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: async () => {
      const { data } = await api.post<AiSummary>(`/ai/case-files/${caseId}/summary`);
      return data;
    },
    onSuccess: () => {
      void qc.invalidateQueries({ queryKey: qk.caseFiles.detail(caseId) });
    },
  });
}

// --- Agentic turn (v4 agent spec §7 / §8.7) ----------------------------------
//
// The single-agent endpoints. All plain axios calls through `lib/api.ts` —
// the SSE/OAC header re-implementation further down stays FROZEN and unused
// by these hooks (the agent turn is buffered by design: a tool loop cannot
// stream usefully, and CloudFront buffers the body anyway).
//
// Types live here (additive) rather than in `api/types.ts`, which this work
// package does not own; the reconciler may fold them in later.

export type AgentActionStatus = "proposed" | "confirmed" | "discarded" | "failed";

/** One @-mention attached to a user message. */
export interface ContextRef {
  entity_type: string;
  entity_id: number;
}

/** Executor output on a confirmed action (`{entity_type, entity_id, url, detail}`). */
export interface AgentActionResult {
  entity_type?: string | null;
  entity_id?: number | null;
  url?: string | null;
  detail?: string | null;
  [key: string]: unknown;
}

/**
 * One proposed/resolved write (rule 6: suggest -> human confirm -> commit).
 * `allowed` is computed per response for the CALLER so the card can render
 * Confirmar disabled-with-reason without a second request; the server still
 * re-checks the real RBAC gate on confirm.
 */
export interface AgentAction {
  id: number;
  thread_id: number;
  message_id: number | null;
  tool: string;
  arguments: Record<string, unknown>;
  summary: string | null;
  module: string;
  action: string;
  status: AgentActionStatus;
  result: AgentActionResult | null;
  error: string | null;
  confirmed_at: string | null;
  created_at: string | null;
  allowed: boolean;
}

/** The linkage stub a `role=tool` message carries in `tool_calls`. */
export interface AgentToolCallMeta {
  tool_call_id?: string | null;
  action_id?: number | null;
  name?: string | null;
  status?: string | null;
}

/** `MessageRead` with the additive v4 fields (tool_calls, context_refs). */
export interface AgentTurnMessage extends AgentMessage {
  tool_calls?: AgentToolCallMeta | Array<Record<string, unknown>> | null;
  context_refs?: ContextRef[] | null;
}

export interface AgentTurnRequest {
  thread_id?: number | null;
  content: string;
  context_refs?: ContextRef[];
}

export interface AgentTurnResponse {
  thread: AgentThread;
  messages: AgentTurnMessage[];
  pending_actions: AgentAction[];
}

/** Local query key for `GET /ai/agent/actions` (keys.ts is not this pass's file). */
export const agentActionsKey = (threadId?: number | null) =>
  ["agent-actions", threadId ?? null] as const;

/** Confirmed writes touch real modules — invalidate what each tool changed. */
const TOOL_INVALIDATIONS: Record<string, ReadonlyArray<readonly string[]>> = {
  create_group: [qk.accountGroups.all, qk.navigator.all],
  attach_client_to_group: [qk.accountGroups.all, qk.navigator.all],
  create_case_file: [qk.caseFiles.all, qk.navigator.all],
  renew_case: [qk.caseFiles.all, qk.navigator.all],
  add_note: [qk.notes.all],
};

/** One agentic turn: context refs + tool loop + pending write proposals. */
export function useAgentTurn() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: async (payload: AgentTurnRequest) => {
      const { data } = await api.post<AgentTurnResponse>("/ai/agent/messages", payload);
      return data;
    },
    onSuccess: (data) => {
      void qc.invalidateQueries({ queryKey: qk.agentThreads.messages(data.thread.id) });
      void qc.invalidateQueries({ queryKey: qk.agentThreads.lists() });
      void qc.invalidateQueries({ queryKey: agentActionsKey(data.thread.id) });
    },
  });
}

/** Rehydration for the page: every action of one (own) thread. */
export function useAgentActions(threadId: number | undefined, status?: AgentActionStatus[]) {
  return useQuery({
    queryKey: [...agentActionsKey(threadId ?? 0), status ?? null] as const,
    enabled: !!threadId,
    queryFn: async () => {
      const params = new URLSearchParams();
      if (threadId) params.set("thread_id", String(threadId));
      for (const s of status ?? []) params.append("status", s);
      const { data } = await api.get<ListResponse<AgentAction>>(
        `/ai/agent/actions?${params.toString()}`,
      );
      return data;
    },
  });
}

/** The COMMIT of rule 6 — executes under the confirming user's real RBAC gate. */
export function useConfirmAgentAction() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: async (actionId: number) => {
      const { data } = await api.post<AgentAction>(`/ai/agent/actions/${actionId}/confirm`);
      return data;
    },
    onSuccess: (action) => {
      void qc.invalidateQueries({ queryKey: agentActionsKey(action.thread_id) });
      void qc.invalidateQueries({ queryKey: qk.agentThreads.messages(action.thread_id) });
      for (const key of TOOL_INVALIDATIONS[action.tool] ?? []) {
        void qc.invalidateQueries({ queryKey: key });
      }
    },
  });
}

/** Declining requires no privilege beyond owning the thread. */
export function useDiscardAgentAction() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: async (actionId: number) => {
      const { data } = await api.post<AgentAction>(`/ai/agent/actions/${actionId}/discard`);
      return data;
    },
    onSuccess: (action) => {
      void qc.invalidateQueries({ queryKey: agentActionsKey(action.thread_id) });
      void qc.invalidateQueries({ queryKey: qk.agentThreads.messages(action.thread_id) });
    },
  });
}

// --- Streaming chat -------------------------------------------------------------

/**
 * SSE reader for `POST /ai/threads/{id}/stream`.
 *
 * It re-implements the two OAC mechanisms `lib/api.ts` owns (dual
 * `Authorization` + `X-Radal-Token`, and `x-amz-content-sha256` over the EXACT
 * body string that is sent) because `fetch` — not axios — is what can read a
 * response body incrementally. `lib/api.ts` itself stays frozen.
 *
 * Behind CloudFront the Function URL is `buffered` by contract, so the whole
 * body lands in one chunk at the end. That is expected — the caller falls back
 * to the non-streaming `/messages` endpoint when no token arrives in time.
 */
export interface StreamCallbacks {
  onStart?: (data: Record<string, unknown>) => void;
  onToken: (delta: string) => void;
  onDone?: (data: Record<string, unknown>) => void;
  onError?: (code: string, detail: string) => void;
}

async function sha256Hex(input: string): Promise<string> {
  const bytes = new TextEncoder().encode(input);
  const digest = await crypto.subtle.digest("SHA-256", bytes);
  return Array.from(new Uint8Array(digest))
    .map((b) => b.toString(16).padStart(2, "0"))
    .join("");
}

export async function streamAgentMessage(
  threadId: number,
  content: string,
  callbacks: StreamCallbacks,
  signal?: AbortSignal,
): Promise<void> {
  const token = getAccessToken();
  const body = JSON.stringify({ content });
  const headers: Record<string, string> = {
    "Content-Type": "application/json",
    Accept: "text/event-stream",
  };
  if (token) {
    headers["Authorization"] = `Bearer ${token}`;
    headers["X-Radal-Token"] = token;
  }
  if (crypto?.subtle) headers["x-amz-content-sha256"] = await sha256Hex(body);

  const response = await fetch(`${API_URL}/ai/threads/${threadId}/stream`, {
    method: "POST",
    headers,
    body,
    signal,
  });

  if (!response.ok || !response.body) {
    const code = response.headers.get("X-Radal-AI-Error") ?? "ai_unavailable";
    let detail = `HTTP ${response.status}`;
    try {
      const payload = (await response.json()) as { detail?: unknown };
      if (typeof payload.detail === "string") detail = payload.detail;
    } catch {
      /* the body was not JSON — keep the status line */
    }
    callbacks.onError?.(code, detail);
    return;
  }

  const reader = response.body.getReader();
  const decoder = new TextDecoder();
  let buffer = "";
  let event = "message";

  const dispatch = (name: string, raw: string) => {
    let data: Record<string, unknown> = {};
    try {
      data = raw ? (JSON.parse(raw) as Record<string, unknown>) : {};
    } catch {
      data = { raw };
    }
    if (name === "start") callbacks.onStart?.(data);
    else if (name === "token") callbacks.onToken(String(data.delta ?? ""));
    else if (name === "done") callbacks.onDone?.(data);
    else if (name === "error") {
      callbacks.onError?.(String(data.code ?? "ai_error"), String(data.detail ?? ""));
    }
  };

  for (;;) {
    const { done, value } = await reader.read();
    if (done) break;
    buffer += decoder.decode(value, { stream: true });

    // SSE frames are separated by a blank line; a frame is `event:` + `data:`.
    let split = buffer.indexOf("\n\n");
    while (split !== -1) {
      const frame = buffer.slice(0, split);
      buffer = buffer.slice(split + 2);
      let payload = "";
      for (const line of frame.split("\n")) {
        if (line.startsWith("event:")) event = line.slice(6).trim();
        else if (line.startsWith("data:")) payload += line.slice(5).trim();
      }
      if (event) dispatch(event, payload);
      event = "message";
      split = buffer.indexOf("\n\n");
    }
  }
}

File: backend/app/schemas/ai.py
"""Pydantic v2 schemas for the AI pipeline (extraction + agent chat).

Two shapes matter here:

``ProposalSuggestion``
    The STANDARD proposal shape the model must emit, and the same shape the
    broker edits before confirming. It is a *suggestion* — never a proposal.
    Nothing in this module writes anything; the commit step takes this payload
    back in (possibly edited) via ``ProposalConfirmRequest``.

``AgentMessage*`` / ``AgentThread*``
    The chat surface over ``agent_thread`` / ``agent_message``.

Money follows the Chilean invariants (docs/v2-architecture.md §4.2):
``net = taxable + exempt``, ``vat = 0.19 * taxable``, ``total = net + vat``,
``comprehensive_rate = taxable_rate + exempt_rate``. The schemas carry every
component as an OPTIONAL field: the model rarely emits all of them and the
service back-fills what it can derive.
"""
from __future__ import annotations

from datetime import date, datetime
from decimal import Decimal
from enum import Enum
from typing import Annotated, Any, Literal

from pydantic import BaseModel, BeforeValidator, ConfigDict, Field, field_validator


def _enum_value(value: Any) -> Any:
    """Unwrap a SQLAlchemy enum member to its lowercase English string value.

    Needed because ``Enum.__hash__`` is the MEMBER NAME, so a ``Literal[...]``
    lookup would miss ``CoverageKind.COVERAGE`` even though it compares equal to
    ``"coverage"``. Applied to every literal alias below so the read schemas
    accept ORM objects directly.
    """
    return value.value if isinstance(value, Enum) else value


EnumStr = Annotated[str, BeforeValidator(_enum_value)]

# --- Shared vocabularies (mirror the model enums, kept as literals so the API
# contract does not leak SQLAlchemy types) ------------------------------------

CoverageKindLiteral = Annotated[
    Literal["coverage", "exclusion"], BeforeValidator(_enum_value)
]
AgentScopeLiteral = Annotated[
    Literal["proposal", "quote", "general"], BeforeValidator(_enum_value)
]
AgentRoleLiteral = Annotated[
    Literal["system", "user", "assistant", "tool"], BeforeValidator(_enum_value)
]
ExtractionStatusLiteral = Annotated[
    Literal["pending", "running", "succeeded", "failed"], BeforeValidator(_enum_value)
]
ProposalOriginLiteral = Annotated[
    Literal["native", "external"], BeforeValidator(_enum_value)
]
ProposalStatusLiteral = Annotated[
    Literal["draft", "submitted", "accepted", "rejected", "withdrawn", "expired"],
    BeforeValidator(_enum_value),
]


class ApiModel(BaseModel):
    """Base: ORM-friendly, trims incidental whitespace on strings."""

    model_config = ConfigDict(from_attributes=True, str_strip_whitespace=True)


# --- Proposal suggestion (the AI output contract) ----------------------------


class DeductibleEntry(ApiModel):
    """One peril's deductible.

    Deliberately permissive (``extra="allow"``): perils differ per insurance
    line and the basis differs per peril — fire is a % of the LOSS, earthquake a
    % of the INSURED AMOUNT of the affected item, with UF minimums. Anything the
    document states beyond these fields is kept verbatim rather than dropped.
    """

    model_config = ConfigDict(from_attributes=True, extra="allow")

    basis: str | None = Field(
        default=None,
        description="loss | insured_amount | fixed | days — what the % applies to",
    )
    pct: Decimal | None = Field(default=None, description="Percentage, 0-100")
    min_uf: Decimal | None = Field(default=None, description="Minimum deductible in UF")
    max_uf: Decimal | None = None
    days: int | None = Field(default=None, description="Waiting period, when time-based")
    detail: str | None = None


class CoverageSuggestion(ApiModel):
    """One covered item or one exclusion, as read off the document."""

    kind: CoverageKindLiteral = "coverage"
    text: str
    normalized_code: str | None = None
    sort_order: int = 0


class InsurerRef(ApiModel):
    """How the extraction names the issuing insurer.

    Matching is by NORMALISED ``cmf_code`` / ``rut`` ONLY — never by name
    (docs §4.3). ``legal_name`` is carried for display and for creating the row
    when no match exists, not for lookup.
    """

    cmf_code: str | None = None
    rut: str | None = None
    legal_name: str | None = None
    trade_name: str | None = None


class ProposalSuggestion(ApiModel):
    """The standard proposal shape — AI-filled, human-editable, not yet committed."""

    insurer: InsurerRef = Field(default_factory=InsurerRef)

    modality: str | None = None
    activity_classification: str | None = None

    # --- Money (UF) ---
    taxable_premium_uf: Decimal | None = None
    exempt_premium_uf: Decimal | None = None
    net_premium_uf: Decimal | None = None
    vat_uf: Decimal | None = None
    total_premium_uf: Decimal | None = None

    # --- Rates (per mille) ---
    taxable_rate_permille: Decimal | None = None
    exempt_rate_permille: Decimal | None = None
    comprehensive_rate_permille: Decimal | None = None

    commission_pct: Decimal | None = None

    validity_business_days: int | None = None
    coverage_start: date | None = None
    coverage_end: date | None = None
    received_at: date | None = None

    deductibles: dict[str, DeductibleEntry] = Field(
        default_factory=dict,
        description='Keyed by peril, e.g. {"fire": {...}, "earthquake": {...}}',
    )
    warranties: str | None = None
    notes: str | None = None

    coverages: list[CoverageSuggestion] = Field(default_factory=list)

    @field_validator("deductibles", mode="before")
    @classmethod
    def _coerce_deductibles(cls, value: Any) -> Any:
        """Accept ``null``, a list of ``{peril: ...}`` rows, or the canonical dict."""
        if value in (None, "", []):
            return {}
        if isinstance(value, list):
            coerced: dict[str, Any] = {}
            for row in value:
                if not isinstance(row, dict):
                    continue
                peril = row.get("peril") or row.get("name") or row.get("kind")
                if peril:
                    coerced[str(peril)] = {
                        k: v for k, v in row.items() if k not in {"peril", "name", "kind"}
                    }
            return coerced
        if isinstance(value, dict):
            return {
                str(k): (v if isinstance(v, dict) else {"detail": str(v)})
                for k, v in value.items()
                if v is not None
            }
        return {}

    @field_validator("coverages", mode="before")
    @classmethod
    def _coerce_coverages(cls, value: Any) -> Any:
        """Accept a list of plain strings as coverages."""
        if value is None:
            return []
        if isinstance(value, list):
            return [{"text": v} if isinstance(v, str) else v for v in value]
        return value


# --- Extraction --------------------------------------------------------------


class ExtractionRequest(ApiModel):
    """Ask the model to read one already-uploaded document."""

    document_id: int = Field(..., gt=0)


class ExtractionRead(ApiModel):
    """The persisted AI job — the audit trail, always returned with the suggestion."""

    id: int
    document_id: int
    kind: EnumStr
    model: str
    prompt_version: str
    status: ExtractionStatusLiteral
    confidence: Decimal | None = None
    error: str | None = None
    prompt_tokens: int | None = None
    completion_tokens: int | None = None
    started_at: datetime | None = None
    finished_at: datetime | None = None
    created_at: datetime | None = None


class ExtractionSuggestionResponse(ApiModel):
    """SUGGEST step: what the model read, plus everything needed to review it."""

    extraction: ExtractionRead
    suggestion: ProposalSuggestion | None = None
    # Raw parsed JSON, kept alongside the typed suggestion so a field the schema
    # could not coerce is still visible to the reviewer instead of vanishing.
    parsed: dict[str, Any] | None = None
    warnings: list[str] = Field(default_factory=list)


# --- Confirm / commit --------------------------------------------------------


class ProposalConfirmRequest(ApiModel):
    """COMMIT step: the reviewed (possibly edited) suggestion becomes a proposal."""

    quote_request_id: int = Field(..., gt=0)
    proposal: ProposalSuggestion
    status: ProposalStatusLiteral = "submitted"
    # The reviewer is asserting the numbers. False only for a save-as-draft flow.
    is_confirmed: bool = True


class ProposalCoverageRead(ApiModel):
    id: int
    kind: CoverageKindLiteral
    text: str
    normalized_code: str | None = None
    sort_order: int


class ProposalRead(ApiModel):
    """The committed proposal."""

    id: int
    broker_id: int
    quote_request_id: int
    insurer_id: int
    insurer_name: str | None = None
    insurer_cmf_code: str | None = None
    origin: ProposalOriginLiteral
    source_document_id: int
    status: ProposalStatusLiteral

    modality: str | None = None
    activity_classification: str | None = None

    taxable_premium_uf: Decimal | None = None
    exempt_premium_uf: Decimal | None = None
    net_premium_uf: Decimal | None = None
    vat_uf: Decimal | None = None
    total_premium_uf: Decimal | None = None
    taxable_rate_permille: Decimal | None = None
    exempt_rate_permille: Decimal | None = None
    comprehensive_rate_permille: Decimal | None = None
    commission_pct: Decimal | None = None

    validity_business_days: int | None = None
    coverage_start: date | None = None
    coverage_end: date | None = None
    received_at: date | None = None

    deductibles: dict[str, Any] | None = None
    warranties: str | None = None
    notes: str | None = None

    extraction_id: int | None = None
    extraction_confidence: Decimal | None = None
    is_confirmed: bool
    confirmed_by_id: int | None = None
    confirmed_at: datetime | None = None

    coverages: list[ProposalCoverageRead] = Field(default_factory=list)


# --- Agent chat --------------------------------------------------------------


class ThreadCreate(ApiModel):
    """Open a conversation. ``proposal``/``quote`` scope anchors it to an entity."""

    scope: AgentScopeLiteral = "general"
    entity_id: int | None = Field(default=None, gt=0)
    title: str | None = Field(default=None, max_length=255)


class ThreadRead(ApiModel):
    id: int
    broker_id: int
    user_id: int
    scope: AgentScopeLiteral
    entity_type: EnumStr | None = None
    entity_id: int | None = None
    title: str | None = None
    is_archived: bool
    last_message_at: datetime | None = None
    created_at: datetime | None = None


class ThreadList(ApiModel):
    items: list[ThreadRead]
    total: int


class MessageCreate(ApiModel):
    content: str = Field(..., min_length=1, max_length=8000)


class MessageRead(ApiModel):
    id: int
    thread_id: int
    role: AgentRoleLiteral
    content: str | None = None
    # Additive (v4 agent): the provider's tool_calls on an assistant message /
    # the linkage stub on a tool message, and the refs a user message carried.
    tool_calls: dict[str, Any] | list[Any] | None = None
    context_refs: list[dict[str, Any]] | None = None
    model: str | None = None
    tokens: int | None = None
    created_at: datetime | None = None


class MessageList(ApiModel):
    items: list[MessageRead]
    total: int


class MessageExchange(ApiModel):
    """One turn: what the user said and what the assistant answered."""

    thread_id: int
    user_message: MessageRead
    assistant_message: MessageRead


# --- Agentic turn (v4 agent spec §7) ------------------------------------------

AgentActionStatusLiteral = Annotated[
    Literal["proposed", "confirmed", "discarded", "failed"],
    BeforeValidator(_enum_value),
]

#: Ref entity types the agent endpoint resolves (spec §6.1). The service is the
#: authority; this mirrors it so the request 422s early with a clear message.
CONTEXT_REF_TYPES = (
    "account_group",
    "client",
    "case_file",
    "policy",
    "quote_request",
    "proposal",
    "document",
    "sales_lead",
)

MAX_CONTEXT_REFS = 6


class ContextRef(ApiModel):
    """One @-mention attached to a user message."""

    entity_type: str = Field(..., min_length=1, max_length=30)
    entity_id: int = Field(..., gt=0)


class AgentTurnRequest(ApiModel):
    """``POST /ai/agent/messages`` — one agentic turn."""

    thread_id: int | None = Field(default=None, gt=0)
    content: str = Field(..., min_length=1, max_length=8000)
    context_refs: list[ContextRef] = Field(default_factory=list, max_length=MAX_CONTEXT_REFS)


class AgentActionRead(ApiModel):
    """One proposed/resolved write. ``allowed`` is computed per response for
    the CALLER, so the card can render Confirmar disabled-with-reason without a
    second request (the server still re-checks on confirm)."""

    id: int
    thread_id: int
    message_id: int | None = None
    tool: str
    arguments: dict[str, Any] = Field(default_factory=dict)
    summary: str | None = None
    module: str
    action: str
    status: AgentActionStatusLiteral
    result: dict[str, Any] | None = None
    error: str | None = None
    confirmed_at: datetime | None = None
    created_at: datetime | None = None
    allowed: bool = False


class AgentActionList(ApiModel):
    items: list[AgentActionRead]
    total: int


class AgentTurnResponse(ApiModel):
    """Everything one agent turn persisted, in order."""

    thread: ThreadRead
    messages: list[MessageRead]
    pending_actions: list[AgentActionRead] = Field(default_factory=list)


# --- Registry-driven extraction (v2) -----------------------------------------
#
# The generic path: any of the 26 registry categories, one response shape, so the
# UI can render a review form for a category it has never seen.


class DocumentExtractionRequest(ApiModel):
    """Ask the model to read one uploaded document with its registry schema."""

    document_id: int = Field(..., gt=0)
    category: str | None = Field(
        default=None,
        description="Override the document's filed category (registry slug)",
    )


# The spec's own name for the same shape.
ExtractionRequestV2 = DocumentExtractionRequest


class CategoryFieldRead(ApiModel):
    """One top-level field of a category schema, for the generic review form."""

    name: str
    type: str | None = None
    required: bool = False
    description: str | None = None


class CategorySpecRead(ApiModel):
    """One registry row as JSON — everything the UI needs to render it."""

    category: str
    canonical_category: str
    is_alias: bool = False
    code: str | None = None
    codes: list[str] = Field(default_factory=list)
    section: str
    direction: str
    prompt_version: str
    module: str
    schema_name: str
    prefill_target: str
    extraction_kind: str
    guidance: str
    fields: list[CategoryFieldRead] = Field(default_factory=list)


class CategoryListResponse(ApiModel):
    items: list[CategorySpecRead]
    total: int


class DocumentExtractionResponse(ApiModel):
    """SUGGEST, generic: the audit row plus the category's own payload."""

    extraction: ExtractionRead
    category: str
    canonical_category: str
    schema_name: str
    schema_version: str
    prefill_target: str
    section: str
    payload: dict[str, Any] | None = None
    # The raw parse, kept alongside the typed payload so a field the schema could
    # not coerce is still visible to the reviewer instead of vanishing.
    parsed: dict[str, Any] | None = None
    # Only filled for the insurer-quotation categories: the legacy contract.
    suggestion: ProposalSuggestion | None = None
    warnings: list[str] = Field(default_factory=list)


class DocumentConfirmRequest(ApiModel):
    """COMMIT the human-reviewed payload for any registry category."""

    payload: dict[str, Any] = Field(default_factory=dict)
    category: str | None = Field(
        default=None, description="Override the extraction's recorded category"
    )
    # Only for the insurer-quotation categories, where confirming really does
    # create a proposal.
    quote_request_id: int | None = Field(default=None, gt=0)
    status: ProposalStatusLiteral = "submitted"
    is_confirmed: bool = True


class DocumentConfirmResponse(ApiModel):
    extraction: ExtractionRead
    category: str
    prefill_target: str
    # True when the confirm wrote the target entity (spec §6.6). False either
    # because the category is genuinely informational (``informational=True``)
    # or because something required is missing (see ``commit.detail``).
    applied: bool = False
    # True for categories that legitimately commit nothing (comparatives,
    # declinations, cover letters, closure notes) — the UI says so instead of
    # looking broken.
    informational: bool = False
    # What the commit did: {target, entity_type, entity_id, detail, ...extras}.
    commit: dict[str, Any] | None = None
    payload: dict[str, Any] | None = None
    proposal: ProposalRead | None = None
    warnings: list[str] = Field(default_factory=list)


class SummaryRead(ApiModel):
    """AI prose, always UNCONFIRMED: suggest -> edit -> confirmar."""

    extraction_id: int
    text: str
    model: str
    prompt_version: str
    is_confirmed: bool = False
    proposal_id: int | None = None
    case_file_id: int | None = None
    pack_id: int | None = None


class StreamFrame(ApiModel):
    """One SSE frame. Documentation-only: the endpoint writes text/event-stream.

    ``start`` -> ``token``* -> ``done``, or a single ``error`` frame.
    """

    event: Literal["start", "token", "done", "error"]
    data: dict[str, Any] = Field(default_factory=dict)


__all__ = [
    "ApiModel",
    "EnumStr",
    "DocumentExtractionRequest",
    "ExtractionRequestV2",
    "CategoryFieldRead",
    "CategorySpecRead",
    "CategoryListResponse",
    "DocumentExtractionResponse",
    "DocumentConfirmRequest",
    "DocumentConfirmResponse",
    "SummaryRead",
    "StreamFrame",
    "DeductibleEntry",
    "CoverageSuggestion",
    "InsurerRef",
    "ProposalSuggestion",
    "ExtractionRequest",
    "ExtractionRead",
    "ExtractionSuggestionResponse",
    "ProposalConfirmRequest",
    "ProposalCoverageRead",
    "ProposalRead",
    "ThreadCreate",
    "ThreadRead",
    "ThreadList",
    "MessageCreate",
    "MessageRead",
    "MessageList",
    "MessageExchange",
    "CONTEXT_REF_TYPES",
    "MAX_CONTEXT_REFS",
    "ContextRef",
    "AgentTurnRequest",
    "AgentActionRead",
    "AgentActionList",
    "AgentTurnResponse",
]
